### PR TITLE
refactor(go): updated go client stub generation

### PR DESCRIPTION
**Pull request by megglos**
## Description

The READMEs for go client related setup was outdated and dependent on a deprecated plugin being [`github.com/golang/protobuf/protoc-gen-go`](https://pkg.go.dev/github.com/golang/protobuf/protoc-gen-go)

This change updates all relevant README files to make use of the new `protoc-gen-go` & `protoc-gen-go-grpc` commands to generate go classes and go-grpc stubs as well as the corresponding maven plugin setup.

### CODE REVIEW

File: clients/go/pkg/worker/exponentialBackoffSupplier_test.go [hunk 1]
```diff
@@ -103,7 +103,7 @@ func TestExponentialBackoffSupplier_ShouldBeRandomizedWithJitter(t *testing.T) {
 	// then
 	for i, delay := range retryDelays {
 		// retryDelay is in bounds
-		betweenBounds := delay > lowerMaxBoundDuration && delay < upperMaxBoundDuration
+		betweenBounds := delay >= lowerMaxBoundDuration && delay <= upperMaxBoundDuration
```
Comment (megglos) [inline clients/go/pkg/worker/exponentialBackoffSupplier_test.go:106]: I hit the edge-case of it being equal 😅 as jitter is computed as `return (e.random.Float64() * (maxFactor - minFactor)) + minFactor`

Comment (ana-vinogradova-camunda) [reply clients/go/pkg/worker/exponentialBackoffSupplier_test.go:106]: Good finding!

**Review by ana-vinogradova-camunda — APPROVED** [review]
Great job, that will help everyone working on Go client.

**Comment by megglos** [comment]
as I'm also a go newbie, @Zelldon could you take a look as well?

File: clients/go/README.md [hunk 2]
```diff
@@ -12,13 +12,7 @@ In order to do this please follow [this guide](../../gateway-protocol-impl/READM
...
 
 ```
-GO111MODULE=off mockgen github.com/camunda/zeebe/clients/go/v8/pkg/pb GatewayClient,Gateway_ActivateJobsClient > internal/mock_pb/mock_gateway.go
-```
-
-If you see errors regarding packages which are not found then as alternative you can try this:
-
-```
-GO111MODULE=off mockgen -source $GOPATH/src/github.com/camunda/zeebe/clients/go/v8/pkg/pb/gateway.pb.go GatewayClient,Gateway_ActivateJobsClient > internal/mock_pb/mock_gateway.go
+mockgen -source=pkg/pb/gateway_grpc.pb.go GatewayClient,Gateway_ActivateJobsClient > internal/mock_pb/mock_gateway.go
```
Comment (Zelldon) [inline clients/go/README.md:15]: 🔧 Before it was using the GOPATH, which was kind of an absolut path, now we using an relative path. Would be good if we then mention that you have to be inside `clients/go` to run this.

Maybe the same what you wrote here https://camunda.slack.com/archives/C05DH1F5TAR/p1702311839690769?thread_ts=1702306903.159889&cid=C05DH1F5TAR

File: clients/go/internal/mock_pb/mock_gateway.go [hunk 3]
```diff
@@ -14,41 +14,6 @@ import (
...
-// EXPECT returns an object that allows the caller to indicate expected use.
-func (m *MockisDeployment_Metadata) EXPECT() *MockisDeployment_MetadataMockRecorder {
-	return m.recorder
-}
-
-// isDeployment_Metadata mocks base method.
-func (m *MockisDeployment_Metadata) isDeployment_Metadata() {
-	m.ctrl.T.Helper()
-	m.ctrl.Call(m, "isDeployment_Metadata")
-}
```
Comment (Zelldon) [inline clients/go/internal/mock_pb/mock_gateway.go:44]: 💭 Wondering why this was removed

File: clients/go/pkg/pb/gateway.pb.go [hunk 4]
```diff
@@ -4716,52 +5077,58 @@ var file_gateway_proto_depIdxs = []int32{
...
-	45, // 56: gateway_protocol.Gateway.UpdateJobRetries:output_type -> gateway_protocol.UpdateJobRetriesResponse
-	49, // 57: gateway_protocol.Gateway.ModifyProcessInstance:output_type -> gateway_protocol.ModifyProcessInstanceResponse
-	51, // 58: gateway_protocol.Gateway.DeleteResource:output_type -> gateway_protocol.DeleteResourceResponse
-	53, // 59: gateway_protocol.Gateway.BroadcastSignal:output_type -> gateway_protocol.BroadcastSignalResponse
-	41, // [41:60] is the sub-list for method output_type
-	22, // [22:41] is the sub-list for method input_type
-	22, // [22:22] is the sub-list for extension type_name
-	22, // [22:22] is the sub-list for extension extendee
-	0,  // [0:22] is the sub-list for field type_name
+	58, // 19: gateway_protocol.ModifyProcessInstanceRequest.activateInstructions:type_name -> gateway_protocol.ModifyProcessInstanceRequest.ActivateInstruction
```
Comment (Zelldon) [inline clients/go/pkg/pb/gateway.pb.go:5080]: :D When was this the last time generated 😅

File: clients/go/pkg/pb/gateway_grpc.pb.go [hunk 5]
```diff
@@ -0,0 +1,1301 @@
+// Code generated by protoc-gen-go-grpc. DO NOT EDIT.
+// versions:
+// - protoc-gen-go-grpc v1.3.0
+// - protoc             v4.25.1
+// source: gateway.proto
+
+package pb
```
Comment (Zelldon) [inline clients/go/pkg/pb/gateway_grpc.pb.go:7]: Interesting 🤔 so we have now two files ? 🤔

File: gateway-protocol-impl/README.md [hunk 6]
```diff
@@ -8,11 +8,12 @@ Zeebe client grpc protobuf protocol.
...
 
-1. Go >= 1.7 (https://golang.org/dl/)
-2. The go protoc generator
+1. Go >= 1.19 (https://golang.org/dl/)
+2. The go & go-grpc protoc generators
 
    ```
-   go get -u github.com/golang/protobuf/protoc-gen-go
+   go install google.golang.org/protobuf/cmd/protoc-gen-go
+   go install google.golang.org/grpc/cmd/protoc-gen-go-grpc
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply gateway-protocol-impl/README.md:16]: Wait. `github.com/golang/protobuf/protoc-gen-go` still exists and should still work. https://github.com/golang/protobuf/issues/1070#issuecomment-607465055

File: gateway-protocol-impl/pom.xml [hunk 7]
```diff
@@ -108,14 +108,23 @@
             <artifactId>protobuf-maven-plugin</artifactId>
             <executions>
               <execution>
+                <id>go</id>
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply gateway-protocol-impl/pom.xml:111]: Afaik I understood the comment is that the plugins have been split of or a separate has been created for the generation of gRPC. But the old program should still work? 🤔

**Review by ChrisKujawa — APPROVED** [review]
Thanks 👍🏼 I run into issues also in the past with the mocking. 

I have some comments maybe you can take a look before merging.

File: clients/go/pkg/pb/gateway.pb.go (hunk 4 above)
Comment (megglos) [reply clients/go/pkg/pb/gateway.pb.go:5080]: I think before the 8.3 release as part of the MT epic, why do you ask?

**Comment by megglos** [comment]
> Wait. github.com/golang/protobuf/protoc-gen-go still exists and should still work. https://github.com/golang/protobuf/issues/1070#issuecomment-607465055

yes it does, while trying the commands I got intrigued to follow-up on the deprecation warning and changed it, however that seems to be too much of a change as of now. I will just fallback to only updating the readme to generate the mocks, who knows what side-effects we may see with the new gen tooling 🤷

**Comment by megglos** [comment]
I will close this and update the mocken command in a new PR

**Pull request closed without merging**